cub/agent/agent_batch_memcpy.cuh

```diff
@@ -96,87 +122,169 @@ __forceinline__ __device__ void LoadVector(const char *ptr, VectorT &data_out)
   constexpr uint32_t bits_per_byte = 8U;
   const uint32_t bit_shift         = offset * bits_per_byte;
 
-  LoadVectorAndFunnelShiftR(aligned_ptr, bit_shift, data_out);
+  // If `ptr` is aligned to four bytes, we can perform a simple uint32_t-aliased load
+  if (offset == 0)
+  {
+    LoadVectorAndFunnelShiftR<true>(aligned_ptr, bit_shift, data_out);
+  }
+  // Otherwise, we need to load extra bytes and perform funnel-shifting
+  else
+  {
+    LoadVectorAndFunnelShiftR<false>(aligned_ptr, bit_shift, data_out);
+  }
```

Comment:
I wonder if there would be any advantage to dispatching to where the offset is statically known. It looks like that would allow `bit_shift` to be known statically as well. 

```suggestion
switch(offset)
case 0: LoadVectorAndFunnelShiftR<0>(...)
case 1: LoadVectorAndFunnelShiftR<1>(...)
case 2: LoadVectorAndFunnelShiftR<2>(...)
case 3: LoadVectorAndFunnelShiftR<3>(...)
```

Reply:
Interesting idea! Preliminary results look like it does not cut above the noise, but I'll do a more thorough run and follow up.

Reply:
I've ran some more benchmarks on this suggestion with code paths that use immediate bit-shift values. Performance remained unchanged for `DeviceMemcpy::Batched`. 

My hypothesis is that we're bottlenecked by the memory subsystem, as I also don't see significant performance changes for some other changes that I'd expect to otherwise positively impact performance.